usenix87/Comm/Omen/Rbsb/rb.1
'\" Revision Level 
'\" Last Delta     01-26-86
.TH RB 1 OMEN
.SH NAME
rb \- X/YMODEM batch file receive
.SH SYNOPSIS
.B rb
[
.B \-1bquv
]
.PP
.B rb
[
.B \-1bcquv
]
.I file
.PP
.B [-][v]rbCOMMAND
.SH DESCRIPTION
.B Rb
uses the XMODEM and YMODEM error correcting protocols to receive
files over a serial port from a variety of programs running under
PC-DOS, CP/M, *nix, and other operating systems.

The first form of
.B rb
(Receive Batch)
receives files with YMODEM batch protocol.
Rb accepts either standard 128 byte sectors or
1024 byte sectors
(YAM
.B k
option).
The user should determine experimentally
the conditions under which use of 1k blocks
actually improves throughput without causing
problems.

Normally, the file contents are converted to
.SM Unix
conventions by stripping carriage returns and all characters
beginning with Control Z (CP/M end of file).
If the raw pathname ends in
".A",
".ARC",
".CCC",
".CL",
".CMD",
".COM",
".CRL",
".DAT",
".DIR",
".EXE",
".O",
".OBJ",
".OVL",
".PAG",
".REL",
".SAV",
".SUB",
".SWP",
".SYS",
".TAR",
".UTL",
".a",
".o",
".tar",
or if the first packet contains
data that suggest a binary file
(parity bits or characters in the range 000 to 006 before the first ^Z),
or if the file mode is transmitted and the 0100000 (most significant)
bit is set, that file will be received in binary mode anyway.
Otherwise,
if the raw pathname ends in .MSG, or .TXT, an existing file will
be appended to rather than replaced.

If extended file information (file length, etc.)
is received,
the file length controls the number of bytes written to
the output dataset,
and the modify time and file mode
(iff non zero)
are set accordingly.

If no extended file information is received,
slashes in the pathname are changed to underscore,
and any trailing period in the pathname is eliminated.
Normally, each file name is converted to lower case
unless it contains one or more lower case letters.

If rb is invoked with stdout and stderr to different datasets,
Verbose is set to 2, causing frame by frame progress reports
to stderr.
This may be disabled with the
.B q
option.

The second form of
.B rb
receives a single
.I file
with XMODEM protocol.
The user must supply the file name to both sending and receiving programs.

In the third form,
.B rb
is invoked as
.B rbCOMMAND
(with an optional leading \- as generated by login(1)).
For each received file,
rb will pipe the file to ``COMMAND filename''
where filename is the name of the transmitted file
with the file contents as standard input.

The file transfer is completed when COMMAND exits with 0 status.
A non zero exit status terminates the transfer.

A typical use for this form is
.I rbrmail
which calls rmail(1)
to post mail to the user specified by the transmitted file name.
For example, sending the file "caf" from a PC-DOS system to
.I rbrmail
on a Unix system
would result in the contents of the DOS file "caf" being mailed to user "caf".

On some
.SM Unix
systems, the login directory must contain a link to
COMMAND as login sets SHELL=rsh which disallows absolute
pathnames.
If invoked with a leading ``v'' rb will report progress to LOGFILE.
The following entry works for
.SM Unix
3.0:
.ce
rbrmail::5:1::/bin:/usr/local/rbrmail
If the SHELL environment variable includes
.I "rsh"
or
.I "rksh"
(restricted shell),
rb will not accept absolute pathnames
or references to a parent directory,
will not modify an existing file, and
removes any files received in error.

.PP
The meanings of the available options are:
.PP
.PD 0
.TP
.B 1
Use file descriptor 1 for ioctls and reads (Unix only).
By default, file descriptor 0 is used.
This option allows
.B rb
to be used with the
.I cu
~$
command.
.TP
.B b
Transfer all files in binary
(tell it like it is)
mode.
This option disables file append.
.TP
.B c
Request 16 bit CRC
(single file transfers default to 8 bit checksum).
.TP
.B q
Quiet suppresses verbosity.
.TP
.B v
.IR Verbose
causes a list of file
names to be appended to
/tmp/rblog .
More v's generate more output.
.TP
.B u
Retain upper case letters in file names unconditionally.
.PD
.SH EXAMPLES
.ne 6
(
.SM Unix
command)
.RS
$
.I rb
.br
rb: ready C
.br
.RE
(Pro-YAM command)
.RS
.I <ALT-2>
.br
YAM Command:
.I "sb *.h *.c"
.br
.RE
.SH SEE ALSO
YMODEM.DOC,
IMP(CP/M),
cu(1),
Professional-YAM manual,
sb(omen),
usq(omen),
undos(omen)

Compile time options required
for various operating systems are described in the
rbsb.c source file.
.SH BUGS
Pathnames are restricted to 127 characters.
In XMODEM single file mode, the pathname given on the command line
is still processed as described above.
The CR/LF to NL translation merely deletes CR\'s;
undos(omen) performs a more intelligent translation.
.SH "VMS VERSION"
Some of the #includes with file names enclosed with angle brackets <>
may need to have the angle brackets changed to "", or vice versa.

The VMS version does not set binary mode according to the incoming
file type.
Non binary file processing consists of stripping all characters beginning
with CPMEOF (^Z).

The VMS version does not set the file time.

VMS sometimes loses incoming characters, resulting in retries
and degradation of throughput.

The VMS C standard i/o package and RMS sometimes interact to modify
file contents unexpectedly.

The VMS version does not support invocation as
.B rbCOMMAND .


usenix89/Comm/Xmodem/rb.c
#define VERSION "rb 0.05 7-19-82"

/*
 * rb.c By Chuck Forsberg
 *
 * A program for Unix which can receive
 *  files from computers running YAM or MODEM.
 *  If no filename is given, YAM batch mode is assumed.
 *
 *  Supports the CRC option or regular checksum.
 *  Received pathnames containing no lowercase letters will be changed to lower
 *  case unless -u option is given.
 *
 *  Unless the -b (binary) option is given, \r is discarded and
 *  ^Z (which is also discarded) acts as end of file.
 *
 *  Any slashes in the pathname are changed to underscore.
 *  If the raw pathname ends in .MSG, .TXT or .DOC, any existing file will
 *  be appended to rather than replaced. Trailing periods are eliminated.
 *
 *  If the raw pathname ends in
 *   .COM .CMD .DAT .O .REL .PAG .CRL .OBJ .SAV
 *   or .?Q* (squeezed file), or if the parity bit of the first or second bytes
 *   of the file is set,
 *   that file will be received in binary mode.
 *
 *
 * a log of activities is appended to "rblog"
 *
 * rb is derived from yam2.c and sb.c
 * rb should use Unix System III buffered input to reduce CPU time.
 *  USG UNIX (3.0) ioctl conventions courtesy  Jeff Martin
 * 	cc -O -DV7  rb.c -o rb		vanilla Unix version 7
 *	cc -O -DUSG rb.c -o rb		USG (3.0) Unix
 *  Unix is a trademark of Western Electric Company
 */

#include <stdio.h>
#include <signal.h>
#include <ctype.h>
#ifdef USG
#include <termio.h>
#include <sys/ioctl.h>
#else
#include <sgtty.h>
#endif

char *substr();
FILE *fout;
FILE *logfile;

#define OK 0
#define FALSE 0
#define TRUE 1
#define ERROR (-1)

/* Ward Christensen / CP/M parameters - Don't change these! */
#define ENQ 005
#define CAN ('X'&037)
#define XOFF ('s'&037)
#define XON ('q'&037)
#define SOH 1
#define STX 2
#define EOT 4
#define ACK 6
#define NAK 025
#define CPMEOF 032
#define WANTCRC 0103	/* send C not NAK to get crc not checksum */
#define TIMEOUT (-2)
#define ERRORMAX 5
#define RETRYMAX 5
#define WCEOT (-10)
#define SECSIZ 128	/* cp/m's Magic Number record size */
#define KSIZE 1024	/* record size with k option */

int Lastrx;
int Crcflg;
int Firstsec;
int Eofseen;		/* indicates cpm eof (^Z) has been received */
int totblocks;		/* total number of blocks received */
int errors;

int Batch;
int MakeLCPathname=TRUE;	/* make received pathname lower case */
int Verbose=FALSE;
int Rxbinary=FALSE;	/* receive all files in bin mode */
int Thisbinary;		/* current file is to be received in bin mode */
int blklen;		/* record length of received packets */
char secbuf[KSIZE];
char linbuf[KSIZE];
int lleft=0;		/* number of characters in linbuf */
int llorig;
char *cdq;		/* pointer for removing chars from linbuf */


#ifdef USG
struct termio oldtty, tty;
#else
struct sgttyb oldtty, tty;
#endif

unsigned updcrc();

main(argc, argv)
char *argv[];
{
	register char *cp;
	register npats=0;
	char **patts;
	int exitcode;

	while (--argc) {
		cp = *++argv;
		if(*cp == '-') {
			while( *++cp) {
				switch(*cp) {
				case 'b':
					Rxbinary=TRUE; break;
				case 'k':
				case 'c':
					Crcflg=TRUE; break;
				case 'u':
					MakeLCPathname=FALSE; break;
				case 'v':
					Verbose=TRUE; break;
				default:
					usage();
				}
			}
		}
		else if( !npats && argc>0) {
			if(argv[0][0]) {
				npats=argc;
				patts=argv;
			}
		}
	}
#ifdef USG
	(void) ioctl(0, TCGETA, &oldtty);
	tty = oldtty;
	tty.c_lflag &= ~(ECHO | ICANON | ISIG); /* No echo, crlf mapping, INTR
	or QUIT chars, no crlf delays,
	no erase/kill processing */
	tty.c_iflag = IGNBRK;	/* Ignore break, disable parity check, no
	stripping, no crnl mapping or ^S^Q */
	tty.c_oflag = 0;	/* Transparent output, no delays, no crlf
	mapping */
	tty.c_cflag &= ~PARENB;	/* Leave baud rate alone, disable parity
	generation and checking */
	tty.c_cflag |= CS8;	/* Set character size = 8 */
	tty.c_cc[VMIN] = 20;	/* Satisfy reads when this many chars in */
	tty.c_cc[VTIME] = 1;	/* ... or in this many tenths of seconds */
	(void) ioctl(0, TCSETA, &tty);
#else
	ioctl(1, TIOCEXCL, 0);
	ioctl(1, TIOCGETP, &oldtty);
	tty = oldtty;
	tty.sg_flags |= RAW;
	tty.sg_flags &= ~ECHO;
	ioctl(1, TIOCSETP, &tty);
#endif
	logfile=fopen("rblog", "a");
	if(wcreceive(npats, patts)==ERROR) {
		exitcode=0200;
		sendline(CAN);
		sendline(CAN);
		sendline(CAN);
		sendline(CAN);
	}
	fclose(logfile);
#ifdef USG
	(void) ioctl(0, TCSBRK, 1);	/* Wait for output to drain */
	(void) ioctl(0, TCFLSH, 1);	/* Flush input queue */
	(void) ioctl(0, TCSETAW, &oldtty);	/* Restore original modes */
	(void) ioctl(0, TCXONC,1);	/* Restart output */
#else
	ioctl(1, TIOCSETP, &oldtty);
	ioctl(1, TIOCNXCL, 0);
#endif
	exit(exitcode);
}

usage()
{
	fprintf(stderr,"%s by Chuck Forsberg\n", VERSION);
	fprintf(stderr,"Usage: rb [-bcuv]\n");
	exit(1);
}

wcreceive(argc, argp)
char **argp;
{
	if(Batch || argc==0) {
		Crcflg=TRUE; fprintf(stderr, "Receiving in batch mode ");
		for(;;) {
			totblocks=0;
			if(wcrxpn(secbuf)== ERROR)
				goto fubar;
			if(secbuf[0]==0)
				return OK;
			if(wcrx(secbuf)==ERROR)
				goto fubar;
		}
	}
	else
		for(; --argc>=0;) {
			totblocks=0;
			if(wcrx(*argp++)==ERROR)
				goto fubar;
		}
	return OK;
fubar:
	sendline(CAN);sendline(CAN);sendline(CAN);
	if(fout)
		fclose(fout);
	return ERROR;
}


/*
 * Fetch a pathname from the other end as a C ctyle ASCIZ string.
 * Length is indeterminate as long as less than blklen
 * a null string represents no more files
 */
wcrxpn(rpn)
char *rpn;	/* receive a pathname */
{
	purgeline();
	Firstsec=TRUE;
	sendline(Crcflg?WANTCRC:NAK);
	if(wcgetsec(rpn, 100) != 0) {
		log( "Pathname fetch failed\n");
		return ERROR;
	}
	sendline(ACK);
	return OK;
}

/*
 * Adapted from CMODEM13.C, written by
 * Jack M. Wierda and Roderick W. Hart
 */

wcrx(name)
char *name;
{
	register int sectnum, sectcurr;
	register char sendchar;
	register char *filemode, *p;

	for(p=name; *p; ++p)		/* change / to _ */
		if( *p == '/')
			*p = '_';
	if ( *--p == '.')		/* zap trailing period */
		*p = 0;
	filemode = "w";
	Thisbinary=Rxbinary;
	if (substr(name, ".COM") || substr(name, ".CRL")
	  || substr(name, ".REL") || substr(name, ".PAG")
	  || substr(name, ".CMD")
	  || substr(name, ".SAV")
	  || substr(name, ".O")
	  || substr(name, ".DAT")
	  || substr(name, ".OBJ")
	  || ((p=substr(name, ".")) && p[2] == 'Q' )
	  || ((p=substr(name, ".")) && p[2] == 'q' ))
		Thisbinary = TRUE;
	if(substr(name, ".TXT")
	  || substr(name, ".MSG")
	  || substr(name, ".DOC"))
		filemode = "a";
	if(MakeLCPathname && !IsAnyLower(name))
		uncaps(name);
	fprintf(logfile,  "Receiving %s %s %s\n",
	  name, Thisbinary?"BIN":"", filemode);
	if ((fout=fopen(name, filemode)) == NULL)
		return ERROR;
	Firstsec=TRUE;sectnum=0; Eofseen=FALSE;
	sendchar=Crcflg?WANTCRC:NAK;

	for(;;) {
/*
		purgeline();
*/
		sendline(sendchar);	/* send it now, we're ready! */
		sectcurr=wcgetsec(secbuf, (sectnum&0177)?50:130);
		if(sectcurr==(sectnum+1 &0377)) {

			if(sectnum==0 && ((secbuf[0]&0200) | (secbuf[1]&0200)))
				Thisbinary++;
			sectnum++;
			if(putsec(secbuf, blklen)==ERROR)
				return ERROR;
			sendchar=ACK;
		}
		else if(sectcurr==(sectnum&0377)) {
			log( "Received dup Sector\n");
			sendchar=ACK;
		}
		else if(sectcurr==WCEOT) {
			sendline(ACK);
			/* don't pad the file any more than it already is */
			if(fclose(fout)==ERROR)
				return ERROR;
			return OK;
		}
		else if(sectcurr==ERROR)
			return ERROR;
		else {
			log( "Sync Error\n");
			return ERROR;
		}
	}
}

/*
 * wcgetsec fetches a Ward Christensen type sector.
 * Returns sector number encountered or ERROR if valid sector not received,
 * or CAN CAN received
 * or WCEOT if eot sector
 * time is timeout for first char, set to 4 seconds thereafter
 ***************** NO ACK IS SENT IF SECTOR IS RECEIVED OK **************
 *    (Caller must do that when he is good and ready to get next sector)
 */

wcgetsec(rxbuf, time)
char *rxbuf;
int time;
{
	register checksum, wcj, firstch;
	register unsigned oldcrc;
	register char *p;
	int sectcurr;

	for(Lastrx=errors=0; errors<RETRYMAX; errors++) {

		if ((firstch=readline(time))==STX) {
			blklen=KSIZE; goto get2;
		}
		if (firstch==SOH) {
			blklen=SECSIZ;
get2:
			sectcurr=readline(1);
			if((sectcurr+readline(1))==255) {
				oldcrc=checksum=0;
				for(p=rxbuf,wcj=blklen; --wcj>=0; ) {
					if((firstch=readline(1)) < 0)
						goto bilge;
					oldcrc=updcrc(firstch, oldcrc);
					checksum += (*p++ = firstch);
				}
				if((firstch=readline(1)) < 0)
					goto bilge;
				if(Crcflg) {
					oldcrc=updcrc(firstch, oldcrc);
					if((firstch=readline(1)) < 0)
						goto bilge;
					oldcrc=updcrc(firstch, oldcrc);
					if(oldcrc)
						log("CRC=0%o\n", oldcrc);
					else {
						Firstsec=FALSE;
						return sectcurr;
					}
				}
				else if(((checksum-firstch)&0377)==0) {
					Firstsec=FALSE;
					return sectcurr;
				}
				else
					log( "Checksum Error\n");
			}
			else
				log( "Sector number garbled\n");
		}
		/* make sure eot really is eot and not just mixmash */
		else if(firstch==EOT && readline(1)==TIMEOUT)
			return WCEOT;
		else if(firstch==CAN) {
			if(Lastrx==CAN) {
				log( "Sender CANcelled\n");
				return ERROR;
			} else {
				Lastrx=CAN;
				continue;
			}
		}
		else if(firstch==TIMEOUT) {
			if (Firstsec)
				goto humbug;
bilge:
			log( "Timeout\n");
		}
		else
			log( "Got 0%o sector header\n", firstch);

humbug:
		Lastrx=0;
		while(readline(1)!=TIMEOUT)
			;
		if(Firstsec)
			sendline(Crcflg?WANTCRC:NAK);
		else {
			time=40; sendline(NAK);
		}
	}
	/* try to stop the bubble machine. */
	sendline(CAN); sendline(CAN); sendline(CAN);
	return ERROR;
}

/*
 * This version of readline is not so well suited for
 * reading many characters.
 *
 * timeout is in tenths of seconds
 */
readline(timeout)
int timeout;
{
	if(--lleft>=0)
		return (*cdq++ & 0377);
	settimeout(timeout);
	if((llorig=lleft=read(1, cdq=linbuf, KSIZE))<1)
		return TIMEOUT;
	alarm(0);
	--lleft;
	return (*cdq++ & 0377);
}


alrm()
{
/* does nothing; actual effect is to give an error return on read */
}

/*
 * timeout is in tenths of seconds
 */
settimeout(timeout)
{
	register int c;

	signal(SIGALRM, alrm);
	if((c = timeout/10)<2)
		c=2;
	alarm(c);
}
purgeline()
{
	lleft=0;
	lseek(1, 0L, 2);
}

/* update CRC */
unsigned updcrc(c, crc)
register c;
register unsigned crc;
{
	register count;

	for(count=8; --count>=0;) {
		if(crc & 0x8000) {
			crc <<= 1;
			crc += (((c<<=1) & 0400)  !=  0);
			crc ^= 0x1021;
		}
		else {
			crc <<= 1;
			crc += (((c<<=1) & 0400)  !=  0);
		}
	}
	return crc;	
}

/* make string s lower case */
uncaps(s)
register char *s;
{
	for( ; *s; ++s)
		if(isupper(*s))
			*s = tolower(*s);
}


/*
 * IsAnyLower returns TRUE if string s has lower case letters.
 */
IsAnyLower(s)
register char *s;
{
	for( ; *s; ++s)
		if (islower(*s))
			return TRUE;
	return FALSE;
}
/*
 * putsec writes the n characters of buf to receive file fout.
 *  If not in binary mode, carriage returns, and all characters
 *  starting with CPMEOF are discarded.
 */
putsec(buf, n)
char *buf;
register n;
{
	register char *p;

	++totblocks;
	if(Thisbinary)
	{
		for (p=buf; --n>=0; )
			putc( *p++, fout);
	}
	else {
		if(Eofseen)
			return OK;
		for (p=buf; --n>=0; ++p ) {
			if( *p == '\r')
				continue;
			if (*p == CPMEOF) {
				Eofseen=TRUE; return OK;
			}
			putc(*p ,fout);
		}
	}
	return OK;
}
sendline(c)
{
	putchar(c);
	fflush(stdout);
}


/*
 * substr(string, token) searches for token in string s
 * returns pointer to token within string if found, NULL otherwise
 */
char *substr(s, t)
register char *s,*t;
{
	register char *ss,*tt;
	/* search for first char of token */
	for(ss=s; *s; s++)
		if(*s == *t)
			/* compare token with substring */
			for(ss=s,tt=t; ;) {
				if(*tt == 0)
					return s;
				if(*ss++ != *tt++)
					break;
			}
	return NULL;
}
log(s,p)
char *s, *p;
{
	fprintf(logfile, "error %d: ", errors);
	fprintf(logfile, s,p);
}
